### PR TITLE
session: fix the problem that lost connection in t ... (#18567)

### LINKED ISSUE #18540: memory_quota: lost connection in the next sql after out of memory quota
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

oom-action=cancel
```
drop table if exists t;
create table t(a int);
insert into t values(1), (-1), (0);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert /*+ memory_quota(1 MB) */ into t select * from t;
insert /*+ memory_quota(2 MB) */ into t select * from t;
insert /*+ memory_quota(14 MB) */ into t select * from t;
insert /*+ memory_quota(15 MB) */ into t select * from t;
select count(*) from t;
```

### 2. What did you expect to see? (Required)
```
mysql> select count(*) from t;
+----------+
| count(*) |
+----------+
|   393216 |
+----------+
1 row in set (3.38 sec)
```
### 3. What did you see instead (Required)
```
mysql> select count(*) from t;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
### 4. Affected version (Required)

```
Release Version: v4.0.2-23-g89e34b707
Edition: Community
Git Commit Hash: 89e34b707fdfeb1bcac5b3ecd13f5614bd2f0291
Git Branch: release-4.0
UTC Build Time: 2020-07-14 06:30:07
GoVersion: go1.13.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18567 to release-4.0

---

### What problem does this PR solve?

Issue Number: close #18540

Problem Summary:
lost connection in the next sql after out of memory quota

### What is changed and how it works?

What's Changed:
Add recover function in `runStmt` to avoid losing connection.

### Check List

Tests

- Manual test
```
drop table if exists t;
create table t(a int);
insert into t values(1), (-1), (0);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert /*+ memory_quota(1 MB) */ into t select * from t;
insert /*+ memory_quota(2 MB) */ into t select * from t;
insert /*+ memory_quota(14 MB) */ into t select * from t;
insert /*+ memory_quota(15 MB) */ into t select * from t;
tidb> select count(*) from t;
+----------+
| count(*) |
+----------+
|   393216 |
+----------+
1 row in set (0.03 sec)
```

### Release note
- Fix lost connection in the next sql after out of memory quota.

### CODE REVIEW

File: session/txn.go [hunk 1]
```diff
@@ -540,17 +540,15 @@ func (s *session) HasDirtyContent(tid int64) bool {
...
-		}
 		s.txn.cleanup()
 	}()
+
 	st := &s.txn
+<<<<<<< HEAD
 	txnSize := st.Transaction.Size()
+=======
+	bufLen := st.stmtBufLen()
+>>>>>>> 596c0eb... session: fix the problem that lost connection in the next sql after out of memory quota (#18567)
```
Comment (qw4990) [inline session/txn.go:551]: Please fix conflicts. @xiaodong-ji

**Comment by bobotu** [comment]
PTAL @XuHuaiyu @lysu

**Pull request merged**